File: packages/node/src/instrument.ts
import type { McpServer } from '@modelcontextprotocol/server';
import type { Sink } from './sinks/types.js';
import { ERROR_KIND_META_KEY, type ToolCallEvent } from './events.js';
import { parseTraceparent } from './trace-context.js';
import { classifyError, isErrorKind } from './error-kind.js';
import { applyRedaction, type RedactionConfig } from './redaction.js';
import {
  extractAndStripIntent,
  injectIntentSchema,
  isIntentCaptureEnabled,
  schemaSupportsInjection,
  type IntentCaptureOption
} from './intent-capture.js';
import { EventBuffer } from './buffer.js';
import { boundedString, MAX_IDENTIFIER_LENGTH } from './bounded.js';

export interface InstrumentOptions {
  /** Required: this server's logical name. Not derivable from the McpServer instance (its `serverInfo` is private), so it's an explicit option. */
  serverName: string;
  serverVersion?: string;
  sinks: Sink[];
  /** Opt-in: capture tool arguments at all. Default false — no arguments are ever recorded unless this is true. */
  captureArguments?: boolean;
  redaction?: RedactionConfig;
  /** Opt-in: inject session_id/agent_id/intent into advertised tool schemas. Default false (off for every tool). */
  intentCapture?: IntentCaptureOption;
  /** Host-supplied user/org identity. The library never infers these itself. */
  resolveIdentity?: (ctx: {
    sessionId?: string;
  }) =>
    | { userId?: string; orgId?: string }
    | undefined
    | Promise<{ userId?: string; orgId?: string } | undefined>;
  bufferSize?: number;
  /** Pass `null` for manual mode: no interval timer, no `beforeExit` listener — flush explicitly via the returned handle's `flush()`. */
  flushIntervalMs?: number | null;
}

/**
 * Returned by `instrument()`. `server` is the same instance passed in
 * (instrumentation mutates it in place); use the handle's `server` rather
 * than your own pre-instrument reference so a future non-mutating
 * implementation doesn't silently drop instrumentation under you.
 */
export interface InstrumentHandle {
  server: McpServer;
  /** Flushes any buffered events immediately. On a request-scoped runtime (e.g. Cloudflare Workers), call this via `ctx.waitUntil(handle.flush())` before returning the response. */
  flush(): Promise<void>;
  /**
   * Shuts instrumentation down: a final `flush()`, then the interval timer
   * is cleared and the `beforeExit` listener removed. Call it when the host
   * disposes the server before the process ends - tests, hot reload, one
   * server per connection - so each `instrument()` call releases what it
   * registered. Idempotent. In manual mode (`flushIntervalMs: null`) there
   * is no timer or listener, so it is equivalent to `flush()`.
   */
  close(): Promise<void>;
}

interface ToolCallContext {
  sessionId?: string;
  http?: unknown;
  mcpReq?: {
    id?: unknown;
    /** The request `_meta`, with the reserved `io.modelcontextprotocol/*` keys lifted into `envelope`. */
    _meta?: Record<string, unknown>;
    envelope?: Record<string, unknown>;
  };
}

interface ToolResultLike {
  content?: Array<{ type: string; text?: string }>;
  isError?: boolean;
  resultType?: unknown;
  _meta?: Record<string, unknown>;
}

/** The part of McpServer's `RegisteredTool` this file uses. Its `annotations` change on `update()`. */
interface RegisteredToolLike {
  annotations?: { readOnlyHint?: unknown; destructiveHint?: unknown };
  update?: (updates: { name?: unknown }) => unknown;
}

const PROTOCOL_VERSION_META_KEY = 'io.modelcontextprotocol/protocolVersion';

/** JSON-RPC's code for an error with no code of its own, as the SDK answers it. */
const INTERNAL_ERROR_CODE = -32603;

function booleanOrNull(value: unknown): boolean | null {
  return typeof value === 'boolean' ? value : null;
}

interface ToolsCallRequest {
  params?: { name?: unknown; arguments?: unknown };
}

type ToolsCallHandler = (request: ToolsCallRequest, ctx: ToolCallContext) => unknown;

/** The intent-capture values a call carried, and its arguments without them. */
type IntentFields = ReturnType<typeof extractAndStripIntent>;

function noIntentFields(args: Record<string, unknown>): IntentFields {
  return { clean: args, session_id: null, agent_id: null, intent: null };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

// TextEncoder rather than Buffer.byteLength: this runs on every tool call,
// and Buffer only exists on Cloudflare Workers behind the nodejs_compat flag.
const encoder = new TextEncoder();

function byteLength(value: unknown): number {
  return encoder.encode(JSON.stringify(value) ?? '').length;
}

/**
 * Proves the recorded arguments survive JSON encoding, or throws.
 *
 * Every sink re-serializes `arguments` on its way out - `JSON.stringify` in
 * console/d1/bigquery, `pg`'s own jsonb encoder in postgres - and a value
 * that cannot be encoded makes that whole `write()` throw. EventBuffer
 * catches the throw, so the server is never affected, but the entire batch
 * is dropped with it, including the unrelated events flushed alongside.
 * Checking here means one bad value costs one event's `arguments`, never a
 * whole flush.
 *
 * Only a `redaction.redactor` or a `redaction.allow` entry can produce such
 * a value: the default type-only markers are always encodable. The extra
 * `JSON.stringify` therefore runs only when argument capture is on, and not
 * at all on the default path, where `value` is null.
 */
function assertSerializable(value: Record<string, unknown> | null): Record<string, unknown> | null {
  if (value === null) return null;
  JSON.stringify(value); // throws on a BigInt, a circular reference, or a throwing toJSON
  return value;
}

function extractErrorMessage(result: ToolResultLike): string | null {
  const text = (result.content ?? [])
    .filter(
      (block): block is { type: 'text'; text: string } =>
        block.type === 'text' && typeof block.text === 'string'
    )
    .map(block => block.text)
    .join(' ');
  if (!text) return null;
  return text.length > 2000 ? text.slice(0, 2000) : text;
}

/**
 * Wraps an `McpServer` so every tool registered through it (after this call)
 * records a `tool_call` event. This is the library's one required call —
 * call it immediately after constructing the server and before registering
 * any tools, since it works by wrapping `registerTool` and the `tools/call`
 * handler McpServer installs on its first registration.
 *
 * Every `tools/call` the server answers is recorded, including ones that
 * never reach a tool callback: an unknown or disabled tool, and arguments
 * that fail the tool's `inputSchema`. `success` follows the result the
 * client receives, so a result McpServer rejects against `outputSchema` is
 * recorded as failed.
 *
 * Nothing the library does around a tool call can change what the client
 * receives. The real handler always runs, its return value is passed through
 * unchanged, and a thrown error is re-thrown unchanged; this wrapper only
 * observes it. Every library-side step (request/response byte counting,
 * `resolveIdentity`, redaction, event construction, the buffer push) is
 * guarded: a failure is logged once per `instrument()` call via
 * `console.error`, then suppressed, and the step falls back to a neutral
 * value (`request_bytes`/`response_bytes` 0, empty identity, `arguments:
 * null`). A failing redactor therefore records `arguments: null`, never the
 * raw arguments. A failure inside a sink is handled separately by
 * EventBuffer, also logged once per sink.
 */
export function instrument(server: McpServer, options: InstrumentOptions): InstrumentHandle {
  const buffer = new EventBuffer({
    sinks: options.sinks,
    bufferSize: options.bufferSize,
    flushIntervalMs: options.flushIntervalMs
  });

  // Same "log once, then suppress" pattern EventBuffer uses per sink, scoped
  // to this instrument() call: one line is enough to surface a broken
  // resolver or redactor, and a line per tool call would drown the host's logs.
  let telemetryWarned = false;
  const warnOnce = (step: string, error: unknown): void => {
    if (telemetryWarned) return;
    telemetryWarned = true;
    console.error(
      `[mcpsignals] ${step} failed; the tool result is unaffected and further telemetry errors from this instrument() call are suppressed:`,
      error
    );
  };
  /** Runs one library-side step; on a throw, logs once and returns `fallback`. */
  const guarded = <T>(step: string, fn: () => T, fallback: T): T => {
    try {
      return fn();
    } catch (error) {
      warnOnce(step, error);
      return fallback;
    }
  };
  /** `guarded` for an async step: a throw or a rejection logs once and yields `fallback`. */
  const guardedAsync = async <T>(step: string, fn: () => Promise<T>, fallback: T): Promise<T> => {
    try {
      return await fn();
    } catch (error) {
      warnOnce(step, error);
      return fallback;
    }
  };

  /**
   * What the tool callback saw, handed from the `registerTool` wrapper to the
   * `tools/call` recorder below. Keyed by the request's `ctx`: McpServer
   * passes its `tools/call` handler's `ctx` straight through to the tool
   * callback, so both layers see the same object.
   */
  const handlerCalls = new WeakMap<object, IntentFields>();
  /** Registered tool name -> whether intent capture injected its schema, and the SDK's tool object. */
  const registeredTools = new Map<string, { canInject: boolean; tool?: RegisteredToolLike }>();

  const originalRegisterTool = server.registerTool.bind(server);

  server.registerTool = ((
    name: string,
    config: Record<string, unknown>,
    cb: (...args: unknown[]) => unknown
  ) => {
    const intentEnabled = isIntentCaptureEnabled(options.intentCapture, name);
    const originalInputSchema = config.inputSchema;
    const canInject = intentEnabled && schemaSupportsInjection(originalInputSchema);
    const wrappedConfig = canInject
      ? { ...config, inputSchema: injectIntentSchema(originalInputSchema) }
      : config;

    /** Strips injected intent-capture keys and leaves them for the recorder. */
    const prepare = (args: Record<string, unknown>, ctx: ToolCallContext) => {
      const fields = canInject ? extractAndStripIntent(args) : noIntentFields(args);
      handlerCalls.set(ctx, fields);
      return fields.clean;
    };

    // The SDK's `createToolExecutor` picks the handler arity from the registered
    // `inputSchema` (same truthiness check as here): with a schema it calls
    // `(args, ctx)`; without one it calls `(ctx)` and never forwards the
    // arguments. The wrapper has to match that arity, otherwise the schema-less
    // handler gets the ctx object in `args` and `undefined` in `ctx`. Branch on
    // the final `wrappedConfig.inputSchema`, since intent capture may have
    // injected a schema into a registration that had none.
    //
    // On the schema-less path the SDK does not validate or pass the arguments,
    // so `arguments` is recorded as an empty-arguments call (`{}`) rather than
    // a payload the handler never sees.
    const wrappedCb = wrappedConfig.inputSchema
      ? async (args: Record<string, unknown>, ctx: ToolCallContext) => cb(prepare(args, ctx), ctx)
      : async (ctx: ToolCallContext) => {
          prepare({}, ctx);
          return cb(ctx);
        };

    // The real `registerTool` overloads are exact per input/output schema shape; a
    // generic wrapper can't preserve that precision through a monkey-patch, so we
    // widen to `any` at this one call site rather than fighting the overload set.
    const tool = (originalRegisterTool as (...args: unknown[]) => unknown)(
      name,
      wrappedConfig,
      wrappedCb
    ) as RegisteredToolLike;
    // Only after the SDK accepted it: a duplicate name throws above and must
    // not replace the first registration. `tool` is read at call time, not
    // now, so `RegisteredTool.update({ annotations })` is honored.
    const registration = { canInject, tool };
    registeredTools.set(name, registration);

    // Follow `update({ name })` the way the SDK does: a new name moves the
    // registration, a null or empty one removes it. `remove()` goes through
    // `update`, so this covers it too.
    let currentName = name;
    const originalUpdate = tool.update;
    if (typeof originalUpdate === 'function') {
      tool.update = (updates: { name?: unknown }) => {
        const result = originalUpdate(updates);
        if (updates?.name !== undefined && updates.name !== currentName) {
          if (registeredTools.get(currentName) === registration) {
            registeredTools.delete(currentName);
          }
          if (typeof updates.name === 'string' && updates.name) {
            registeredTools.set(updates.name, registration);
            currentName = updates.name;
          }
        }
        return result;
      };
    }
    return tool;
  }) as typeof server.registerTool;

  /**
   * Runs one `tools/call` request with telemetry around it. This wraps the
   * handler McpServer registers, not the tool callback, so it sees what the
   * client sees: an unknown or disabled tool (a JSON-RPC error), an
   * input-schema failure (an `isError` result, callback never runs), and an
   * output-schema failure (an `isError` result, callback already returned
   * success). A tool callback that throws reaches this layer as McpServer's
   * `isError` result carrying the error message.
   */
  const observeToolsCall = async (
    request: ToolsCallRequest,
    ctx: ToolCallContext,
    handler: ToolsCallHandler
  ) => {
    const startedAt = new Date();
    const start = performance.now();

    const params = request?.params ?? {};
    const rawName = typeof params.name === 'string' ? params.name : '';
    const rawArgs = params.arguments ?? {};
    // `JSON.stringify` throws on a BigInt (and on a throwing `toJSON`);
    // an unmeasurable request is recorded as 0 bytes, never blocks the handler.
    const requestBytes = guarded('request byte count', () => byteLength(rawArgs), 0);

    /**
     * Builds and pushes the event. Every caller wraps this in `guardedAsync`,
     * so a failure anywhere in here (identity, client info, redaction, the
     * buffer push) is logged once and the event is dropped, never surfaced
     * to the client. Redaction and `resolveIdentity` are guarded on their
     * own so a broken redactor or resolver still leaves an event behind,
     * recorded with `arguments: null` / a null identity.
     *
     * `durationMs` is measured by the caller the moment the handler settles.
     * `resolveIdentity` runs in here, after the handler, so its latency is
     * outside the timed window: `duration_ms` is wall time from call start
     * to response, per schema/events.md.
     */
    const push = async (
      durationMs: number,
      partial: Pick<
        ToolCallEvent,
        'success' | 'error_kind' | 'error_message' | 'response_bytes' | 'result_type' | 'error_code'
      >
    ) => {
      const registration = registeredTools.get(rawName);
      // The callback's view when it ran. When it never ran, read the raw
      // arguments the way it would have, so a rejected call still records intent.
      const rawRecord = isRecord(rawArgs) ? rawArgs : {};
      const fields =
        handlerCalls.get(ctx) ??
        (registration?.canInject ? extractAndStripIntent(rawRecord) : noIntentFields(rawRecord));

      // Revision 2026-07-28 carries the version on every request's envelope;
      // earlier revisions negotiate it once in `initialize`. Either way the
      // client supplied it, so it takes the identifier cap. The accessor is
      // deprecated for the same reason as getClientVersion() below and is
      // only the fallback for a request with no envelope.
      const protocolVersion = boundedString(
        ctx.mcpReq?.envelope?.[PROTOCOL_VERSION_META_KEY] ??
          server.server.getNegotiatedProtocolVersion(),
        MAX_IDENTIFIER_LENGTH
      );
      const requestId = ctx.mcpReq?.id;
      // `_meta` is the MCP protocol's field name.
      // oxlint-disable-next-line no-underscore-dangle
      const trace = parseTraceparent(ctx.mcpReq?._meta?.traceparent);
      const annotations = registration?.tool?.annotations;

      // A host resolver that throws or rejects records a null identity.
      let identity: { userId?: string; orgId?: string } = {};
      try {
        identity = (await options.resolveIdentity?.({ sessionId: ctx.sessionId })) ?? {};
      } catch (error) {
        warnOnce('resolveIdentity', error);
      }
      // getClientVersion() is deprecated in favor of reading client identity off the
      // per-request `_meta` envelope, but the SDK's own deprecation note says the accessor
      // "remains functional" and is backfilled per request on 2026-07-28-era connections too.
      // Deliberately kept rather than reaching into the envelope's internal shape, which isn't
      // part of this SDK's stable public surface yet.
      //
      // The client declares its own name/version in the `initialize` handshake, so
      // both are caller-controlled and take the same identifier cap as the intent
      // fields (see bounded.ts).
      const clientInfo = server.server.getClientVersion();
      const clientName = boundedString(clientInfo?.name, MAX_IDENTIFIER_LENGTH);
      const clientVersion = boundedString(clientInfo?.version, MAX_IDENTIFIER_LENGTH);
      const recordedArguments = guarded(
        'redaction',
        () =>
          assertSerializable(
            applyRedaction(fields.clean, options.redaction, options.captureArguments)
          ),
        null
      );
      buffer.push({
        event_type: 'tool_call',
        ts: startedAt,
        server_name: options.serverName,
        server_version: options.serverVersion ?? null,
        // A registered name is the server's own. Any other name came from the
        // client, so it takes the identifier cap.
        tool_name: registeredTools.has(rawName)
          ? rawName
          : (boundedString(rawName, MAX_IDENTIFIER_LENGTH) ?? ''),
        session_id: ctx.sessionId ?? fields.session_id,
        agent_id: fields.agent_id,
        client_name: clientName,
        client_version: clientVersion,
        user_id: identity.userId ?? null,
        org_id: identity.orgId ?? null,
        duration_ms: durationMs,
        request_bytes: requestBytes,
        arguments: recordedArguments,
        intent: fields.intent,
        transport: ctx.http ? 'http' : 'stdio',
        protocol_version: protocolVersion,
        request_id:
          typeof requestId === 'string' || typeof requestId === 'number'
            ? boundedString(String(requestId), MAX_IDENTIFIER_LENGTH)
            : null,
        trace_id: trace?.traceId ?? null,
        parent_span_id: trace?.parentSpanId ?? null,
        read_only_hint: booleanOrNull(annotations?.readOnlyHint),
        destructive_hint: booleanOrNull(annotations?.destructiveHint),
        ...partial
      });
    };

    // Only the handler call itself lives in this try: a throw here is a
    // protocol error the client receives as a JSON-RPC error, recorded and
    // re-thrown unchanged. Recording happens outside it so a telemetry
    // failure can never be mistaken for one.
    let result: ToolResultLike;
    try {
      result = (await handler(request, ctx)) as ToolResultLike;
    } catch (error) {
      const durationMs = Math.round(performance.now() - start);
      await guardedAsync(
        'event recording',
        async () => {
          const message = error instanceof Error ? error.message : String(error);
          const truncated = message.length > 2000 ? message.slice(0, 2000) : message;
          // The same rule the SDK uses to answer: the error's own integer
          // `code` (a ProtocolError's), otherwise internal error.
          const code = (error as { code?: unknown } | null)?.code;
          await push(durationMs, {
            success: false,
            error_kind: classifyError(truncated),
            error_message: truncated,
            response_bytes: 0,
            result_type: null,
            error_code: Number.isSafeInteger(code) ? (code as number) : INTERNAL_ERROR_CODE
          });
        },
        undefined
      );
      throw error;
    }

    const durationMs = Math.round(performance.now() - start);
    await guardedAsync(
      'event recording',
      async () => {
        const responseBytes = guarded('response byte count', () => byteLength(result), 0);
        // Absent means complete: the SDK stamps `resultType` after this layer.
        const resultType = result?.resultType === 'input_required' ? 'input_required' : 'complete';
        if (result?.isError) {
          const errorMessage = extractErrorMessage(result);
          // A kind the handler declared wins; an unknown value falls back to the heuristic.
          // Guarded on its own so a throwing `_meta` costs only the declared kind, not the event.
          const declaredKind = guarded(
            'declared error kind',
            // `_meta` is the MCP protocol's field name.
            // oxlint-disable-next-line no-underscore-dangle
            () => result._meta?.[ERROR_KIND_META_KEY],
            undefined
          );
          await push(durationMs, {
            success: false,
            error_kind: isErrorKind(declaredKind) ? declaredKind : classifyError(errorMessage),
            error_message: errorMessage,
            response_bytes: responseBytes,
            result_type: resultType,
            error_code: null
          });
        } else {
          await push(durationMs, {
            success: true,
            error_kind: null,
            error_message: null,
            response_bytes: responseBytes,
            result_type: resultType,
            error_code: null
          });
        }
      },
      undefined
    );
    return result;
  };

  // McpServer installs its `tools/call` handler through the public
  // `Server.setRequestHandler` the first time a tool is registered, which is
  // after this call. Wrap that one registration; every other method passes
  // through untouched.
  const lowLevel = server.server;
  const originalSetRequestHandler = lowLevel.setRequestHandler.bind(lowLevel) as (
    ...args: unknown[]
  ) => unknown;
  lowLevel.setRequestHandler = ((method: unknown, ...rest: unknown[]) => {
    const [handler] = rest;
    if (method !== 'tools/call' || rest.length !== 1 || typeof handler !== 'function') {
      return originalSetRequestHandler(method, ...rest);
    }
    return originalSetRequestHandler(method, (request: ToolsCallRequest, ctx: ToolCallContext) =>
      observeToolsCall(request, ctx, handler as ToolsCallHandler)
    );
  }) as typeof lowLevel.setRequestHandler;

  return {
    server,
    flush: () => buffer.flush(),
    close: async () => {
      try {
        await buffer.flush();
      } finally {
        buffer.stop();
      }
    }
  };
}

File: packages/go/instrument.go
package mcpsignals

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"strings"
	"time"

	"github.com/modelcontextprotocol/go-sdk/auth"
	"github.com/modelcontextprotocol/go-sdk/mcp"
)

// Identity is supplied only by the application. Empty fields become null.
type Identity struct {
	UserID string
	OrgID  string
}

// CallContext contains detached metadata for an application identity resolver.
// No user identity is inferred from client metadata, arguments or headers.
type CallContext struct {
	SessionID     string
	ClientName    string
	ClientVersion string
	Transport     string
	// TokenInfo is this request's verified bearer token, set by the SDK's
	// auth.RequireBearerToken on streamable HTTP. Nil otherwise. Read-only.
	TokenInfo *auth.TokenInfo
	// Header is a copy of this request's HTTP headers, nil for other transports.
	Header http.Header
}

// Options configures instrumentation. Supply the same name/version passed to
// mcp.NewServer: the SDK does not expose the server's Implementation publicly.
// Configure once, before connecting clients. Callbacks must be concurrency-safe.
type Options struct {
	ServerName       string
	ServerVersion    string
	Buffer           BufferOptions
	CaptureArguments bool
	Redaction        Redaction
	ResolveIdentity  func(context.Context, CallContext) (Identity, error)
	// Transport is "stdio", "http", or empty. A request with HTTP headers
	// (streamable HTTP) always records "http". Otherwise this value is used, and
	// empty means "stdio", as in Node/Python. Set "http" for the legacy SSE
	// transport, which exposes no headers to middleware.
	Transport string
}

// Handle owns the telemetry lifecycle, not the MCP server or its sinks.
type Handle struct {
	Server  *mcp.Server
	buffer  *EventBuffer
	options Options
}

// Instrument adds supported receiving middleware without changing tool handlers
// or registration APIs. Existing and subsequently registered tools are observed.
// Call once per server, before serving clients. Middleware added earlier runs
// inside this one, so ResolveIdentity cannot see context values it adds.
func Instrument(server *mcp.Server, o Options) (*Handle, error) {
	if server == nil || o.ServerName == "" {
		return nil, errors.New("mcpsignals: server and server name are required")
	}
	if o.Transport != "" && o.Transport != "stdio" && o.Transport != "http" {
		return nil, errors.New("mcpsignals: transport must be stdio, http, or empty")
	}
	o.Redaction.Allow = append([]string(nil), o.Redaction.Allow...)
	o.Redaction.Deny = append([]string(nil), o.Redaction.Deny...)
	b, err := NewEventBuffer(o.Buffer)
	if err != nil {
		return nil, err
	}
	h := &Handle{Server: server, buffer: b, options: o}
	server.AddReceivingMiddleware(h.middleware)
	return h, nil
}

// Flush waits for previously started writes and flushes queued events.
func (h *Handle) Flush(ctx context.Context) error { return h.buffer.Flush(ctx) }

// Close drains telemetry after the application has stopped and drained requests.
func (h *Handle) Close(ctx context.Context) error { return h.buffer.Close(ctx) }

// Stats returns buffer overflow and sink failure counters.
func (h *Handle) Stats() BufferStats { return h.buffer.Stats() }

func (h *Handle) middleware(next mcp.MethodHandler) mcp.MethodHandler {
	return func(ctx context.Context, method string, req mcp.Request) (mcp.Result, error) {
		call, ok := req.(*mcp.CallToolRequest)
		if method != "tools/call" || !ok || call == nil || call.Params == nil {
			return next(ctx, method, req)
		}
		start := time.Now()
		// A handler may replace Params fields; record what the call arrived with.
		name, args := call.Params.Name, call.Params.Arguments
		// Deliberately no recovery around next: handler errors/panics belong to the SDK.
		result, err := next(ctx, method, req)
		duration := time.Since(start)
		if r, ok := result.(*mcp.CallToolResult); ok && r != nil && r.NeedsInput() {
			// Not a completed invocation: the client answers the input requests
			// and retries, and that continuation is recorded.
			return result, err
		}
		// Identity and redaction run after the handler, as in Node/Python, so a
		// slow resolver neither delays the handler nor shifts ts.
		event := h.prepare(ctx, call, name, args)
		event.TS = start.UTC()
		event.DurationMS = durationMS(duration)
		h.record(event, result, err)
		return result, err
	}
}

func (h *Handle) prepare(ctx context.Context, req *mcp.CallToolRequest, name string, args json.RawMessage) (e ToolCallEvent) {
	e = ToolCallEvent{EventType: "tool_call", ServerName: h.options.ServerName, ServerVersion: optional(h.options.ServerVersion), ToolName: truncate(name, 128), RequestBytes: len(args)}
	// Telemetry never changes the invocation, even if SDK metadata access fails.
	defer func() { _ = recover() }()
	e.Arguments = captureArguments(args, h.options.CaptureArguments, h.options.Redaction)
	c := CallContext{Transport: h.options.Transport}
	if req.Extra != nil {
		c.TokenInfo = req.Extra.TokenInfo
		if req.Extra.Header != nil {
			c.Transport = "http"
			if h.options.ResolveIdentity != nil {
				c.Header = req.Extra.Header.Clone()
			}
		}
	}
	if c.Transport == "" {
		c.Transport = "stdio"
	}
	if req.Session != nil {
		c.SessionID = req.Session.ID()
	}
	if info := req.ClientInfo(); info != nil {
		c.ClientName = info.Name
		c.ClientVersion = info.Version
	}
	e.Transport = optional(c.Transport)
	e.SessionID = optional(c.SessionID)
	e.ClientName = bounded(c.ClientName, 128)
	e.ClientVersion = bounded(c.ClientVersion, 128)
	// A resolver error or panic leaves user_id and org_id null.
	if h.options.ResolveIdentity != nil {
		if identity, err := h.options.ResolveIdentity(ctx, c); err == nil {
			e.UserID = optional(identity.UserID)
			e.OrgID = optional(identity.OrgID)
		}
	}
	return e
}

func (h *Handle) record(e ToolCallEvent, result mcp.Result, callErr error) {
	setOutcome(&e, result, callErr)
	h.buffer.Push(e)
}

// setOutcome fills the outcome fields. A panic keeps the fields set so far,
// so the call is still recorded.
func setOutcome(e *ToolCallEvent, result mcp.Result, callErr error) {
	defer func() { _ = recover() }()
	e.Success = callErr == nil
	var message string
	var declared string
	if callErr != nil {
		message = callErr.Error()
	} else if r, ok := result.(*mcp.CallToolResult); ok && r != nil {
		e.ResponseBytes = serializedSize(r)
		e.Success = !r.IsError
		if r.IsError {
			var parts []string
			for _, content := range r.Content {
				if text, ok := content.(*mcp.TextContent); ok && text != nil {
					parts = append(parts, text.Text)
				}
			}
			message = strings.Join(parts, " ")
			switch value := r.Meta[ErrorKindMetaKey].(type) {
			case string:
				declared = value
			case ErrorKind:
				declared = string(value)
			}
		}
	}
	if !e.Success {
		message = truncate(message, 2000)
		e.ErrorMessage = optional(message)
		kind := ClassifyError(message)
		if IsErrorKind(declared) {
			kind = ErrorKind(declared)
		}
		if kind != "" {
			e.ErrorKind = &kind
		}
	}
}

// durationMS rounds to the nearest millisecond, like Node's Math.round, so
// rows from every package agree for sub-millisecond calls.
func durationMS(d time.Duration) int64 { return d.Round(time.Millisecond).Milliseconds() }

func serializedSize(v any) (size int) {
	defer func() { _ = recover() }()
	data, err := json.Marshal(v)
	if err == nil {
		return len(data)
	}
	return 0
}
